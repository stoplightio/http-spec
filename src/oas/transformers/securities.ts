import { HttpOperationSecurityDeclarationTypes, IHttpOperation } from '@stoplight/types';
import type { OperationObject } from 'openapi3-ts/src/model/OpenApi';
import type { Operation } from 'swagger-schema-official';

export function translateToSecurityDeclarationType({
  security,
}: Partial<Operation | OperationObject>): IHttpOperation['securityDeclarationType'] {
  let securityDeclarationType = HttpOperationSecurityDeclarationTypes.InheritedFromService;

  if (Array.isArray(security)) {
    securityDeclarationType =
      security.length === 0
        ? HttpOperationSecurityDeclarationTypes.None
        : HttpOperationSecurityDeclarationTypes.Declared;
  }

  return securityDeclarationType;
}
