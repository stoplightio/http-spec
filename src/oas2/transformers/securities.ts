import { isPlainObject } from '@stoplight/json';
import type {
  DeepPartial,
  IApiKeySecurityScheme,
  IBasicSecurityScheme,
  IOauth2SecurityScheme,
  IOauthFlowObjects,
  Optional,
} from '@stoplight/types';
import pickBy = require('lodash.pickby');
import type {
  ApiKeySecurity,
  BasicAuthenticationSecurity,
  OAuth2AccessCodeSecurity,
  OAuth2ApplicationSecurity,
  OAuth2ImplicitSecurity,
  OAuth2PasswordSecurity,
} from 'swagger-schema-official';

import { isNonNullable, isString } from '../../guards';
import { SecurityWithKey } from '../../oas3/accessors';
import { getSecurities } from '../accessors';
import { isSecurityScheme } from '../guards';
import { Oas2TranslateFunction } from '../types';

export const translateToFlows: Oas2TranslateFunction<[security: Record<string, unknown>], IOauthFlowObjects> =
  function (security) {
    const flows: IOauthFlowObjects = {};

    const scopes = isPlainObject(security.scopes) ? pickBy<unknown, string>(security.scopes, isString) : {};
    const authorizationUrl =
      'authorizationUrl' in security && typeof security.authorizationUrl === 'string' ? security.authorizationUrl : '';
    const tokenUrl = 'tokenUrl' in security && typeof security.tokenUrl === 'string' ? security.tokenUrl : '';

    if (security.flow === 'implicit') {
      flows.implicit = {
        authorizationUrl,
        scopes,
      };
    } else if (security.flow === 'password') {
      flows.password = {
        tokenUrl,
        scopes,
      };
    } else if (security.flow === 'application') {
      flows.clientCredentials = {
        tokenUrl,
        scopes,
      };
    } else if (security.flow === 'accessCode') {
      flows.authorizationCode = {
        authorizationUrl,
        tokenUrl,
        scopes,
      };
    }

    return flows;
  };

export const translateToBasicSecurityScheme: Oas2TranslateFunction<
  [security: DeepPartial<BasicAuthenticationSecurity> & { key: string }],
  IBasicSecurityScheme
> = function (security) {
  return {
    type: 'http',
    scheme: 'basic',
    description: security.description,
    key: security.key,
  };
};

const ACCEPTABLE_SECURITY_ORIGINS: ApiKeySecurity['in'][] = ['query', 'header'];

export const translateToApiKeySecurityScheme: Oas2TranslateFunction<
  [security: DeepPartial<ApiKeySecurity> & { key: string }],
  Optional<IApiKeySecurityScheme>
> = function (security) {
  if ('in' in security && security.in && ACCEPTABLE_SECURITY_ORIGINS.includes(security.in)) {
    return {
      type: 'apiKey',
      in: security.in,
      name: String(security.name || ''),
      description: security.description,
      key: security.key,
    };
  }

  return;
};

const VALID_OAUTH2_FLOWS = ['implicit', 'password', 'application', 'accessCode'];

export const translateToOauth2SecurityScheme: Oas2TranslateFunction<
  [
    security: DeepPartial<
      OAuth2AccessCodeSecurity | OAuth2ApplicationSecurity | OAuth2ImplicitSecurity | OAuth2PasswordSecurity
    > & { key: string },
  ],
  Optional<IOauth2SecurityScheme>
> = function (security) {
  if (!security.flow || !VALID_OAUTH2_FLOWS.includes(security.flow)) return undefined;

  return {
    type: 'oauth2',
    flows: translateToFlows.call(this, security),
    description: security.description,
    key: security.key,
  };
};

export const translateToSingleSecurity: Oas2TranslateFunction<
  [security: unknown & { key: string }],
  Optional<SecurityWithKey>
> = function (security) {
  if (isSecurityScheme(security)) {
    switch (security.type) {
      case 'basic':
        return translateToBasicSecurityScheme.call(this, security);
      case 'apiKey':
        return translateToApiKeySecurityScheme.call(this, security);
      case 'oauth2':
        return translateToOauth2SecurityScheme.call(this, security);
    }
  }

  return;
};

export const translateToSecurities: Oas2TranslateFunction<[operationSecurities: unknown], SecurityWithKey[][]> =
  function (operationSecurities) {
    const securities = getSecurities(this.document, operationSecurities);

    return securities.map(security => security.map(translateToSingleSecurity, this).filter(isNonNullable));
  };
