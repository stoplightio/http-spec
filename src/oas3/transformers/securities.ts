import { DeepPartial, HttpSecurityScheme, IOauthFlowObjects } from '@stoplight/types';
import { pickBy } from 'lodash';
import { OAuthFlowsObject, SecuritySchemeObject } from 'openapi3-ts';

export function translateToSecurities(securities: SecuritySchemeObject[][]): HttpSecurityScheme[][] {
  return securities.map(security => {
    return security.map(transformToSingleSecurity);
  });
}

export function transformToSingleSecurity(securityScheme: DeepPartial<SecuritySchemeObject>): HttpSecurityScheme {
  const baseObject: HttpSecurityScheme = {
    name: securityScheme.name as string,
    description: securityScheme.description,
    in: securityScheme.in as 'query' | 'header' | 'cookie',
    type: securityScheme.type as any,
  };

  if (securityScheme.flows) {
    Object.assign(baseObject, {
      flows: (securityScheme.flows && transformFlows(securityScheme.flows)) || {},
    });
  }

  if (securityScheme.openIdConnectUrl) {
    Object.assign(baseObject, { openIdConnectUrl: securityScheme.openIdConnectUrl });
  }

  if (securityScheme.scheme) {
    Object.assign(baseObject, { scheme: securityScheme.scheme });
  }

  if (securityScheme.bearerFormat) {
    Object.assign(baseObject, { bearerFormat: securityScheme.bearerFormat });
  }

  return baseObject;
}

function transformFlows(flows: DeepPartial<OAuthFlowsObject>): IOauthFlowObjects {
  const transformedFlows: IOauthFlowObjects = {};

  if (flows.password) {
    Object.assign(transformedFlows, {
      password: pickBy({
        refreshUrl: flows.password.refreshUrl,
        scopes: flows.password.scopes || {},
        tokenUrl: flows.password.tokenUrl || '',
      }),
    });
  }

  if (flows.implicit) {
    Object.assign(transformedFlows, {
      implicit: pickBy({
        authorizationUrl: flows.implicit.authorizationUrl || '',
        refreshUrl: flows.implicit.refreshUrl,
        scopes: flows.implicit.scopes || {},
      }),
    });
  }

  if (flows.authorizationCode) {
    Object.assign(transformedFlows, {
      authorizationCode: pickBy({
        authorizationUrl: flows.authorizationCode.authorizationUrl || '',
        refreshUrl: flows.authorizationCode.refreshUrl,
        scopes: flows.authorizationCode.scopes || {},
        tokenUrl: flows.authorizationCode.tokenUrl || '',
      }),
    });
  }

  if (flows.clientCredentials) {
    Object.assign(transformedFlows, {
      clientCredentials: pickBy({
        tokenUrl: flows.clientCredentials.tokenUrl || '',
        refreshUrl: flows.clientCredentials.refreshUrl,
        scopes: flows.clientCredentials.scopes || {},
      }),
    });
  }

  return transformedFlows;
}
