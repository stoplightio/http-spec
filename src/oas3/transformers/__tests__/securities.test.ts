import { SecuritySchemeObject } from 'openapi3-ts';

import { translateToSecurities } from '../securities';

describe('translateOas3ToSecurities', () => {
  test('basic snapshot', () => {
    const collection: [SecuritySchemeObject[]] = [
      [
        {
          bearerFormat: 'Bearer',
          description: 'Description',
          in: 'header',
          name: 'oauth2',
          openIdConnectUrl: 'https://oaid.com/login',
          scheme: 'scheme',
          type: 'oauth2',
          flows: {
            authorizationCode: {
              tokenUrl: 'https://google.it/token',
              scopes: {
                'read:write': {},
              },
            },
          },
        },
      ],
    ];
    expect(translateToSecurities(collection)).toMatchSnapshot();
  });
});
