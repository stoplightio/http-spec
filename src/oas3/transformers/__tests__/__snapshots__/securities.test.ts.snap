// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateOas3ToSecurities basic snapshot 1`] = `
Array [
  Array [
    Object {
      "bearerFormat": "Bearer",
      "description": "Description",
      "flows": Object {
        "authorizationCode": Object {
          "scopes": Object {
            "read:write": Object {},
          },
          "tokenUrl": "https://google.it/token",
        },
      },
      "in": "header",
      "name": "oauth2",
      "openIdConnectUrl": "https://oaid.com/login",
      "scheme": "scheme",
      "type": "oauth2",
    },
  ],
]
`;
